### ISSUE TITLE
Bug with dynamic parameters

### ISSUE THREAD

**Issue by vladtar** [ui-router/sticky-states#13]
Consider the following (somewhat complicated) scenario:
```
{
	name: 'stateA'
},
{
	name: 'stateB',
	sticky: true,
	url: '/stateB?paramB',
	params: {
		paramB: {
			dynamic: true
		}
	}
}
```
Let's assume `stateB` is inactive and `paramB` is equal to 1. The current url is `/stateB/?paramB=1`.  `stateA` is currently active. If at this moment we do `$state.go('stateB', { paramB: 2 });` the url does not get updated to `/stateB/?paramB=2`, it stays at `/stateB/?paramB=1`.

Digging into it a bit, in `@ui-router/sticky-states` https://github.com/ui-router/sticky-states/blob/master/src/stickyStates.ts#L197:
```
let simulatedTC = PathUtils.treeChanges(inactiveFromPath, tc.to, trans.options().reloadState);
```

This calls into `@ui-router/core`, the method above returns https://github.com/ui-router/core/blob/master/src/path/pathFactory.ts#L110-L138, specifically:
```
 // Create a new retained path (with shallow copies of nodes) which have the params of the toPath mapped
let retainedWithToParams  = retained.map(applyToParams);
entering              = toPath.slice(keep);
to                    = (retainedWithToParams).concat(entering);

return { from, to, retained, exiting, entering };
```
This method returns the original `retained`, not `retainedWithToParams`.

Back to `@ui-router/sticky-states` this uses https://github.com/ui-router/sticky-states/blob/master/src/stickyStates.ts#L204:
```
tc.reactivating = simulatedTC.retained.slice(tc.retained.length);
```
So reactivating here ends up being the inactive state but with still the old parameters, not the new ones.

There are probably a few different ways to fix this, not sure what would be the preferred one.

Please let me know if this makes sense or whether there is any additional information I should submit.

Thank you.

**Comment by kuhnroyal** [comment]
I just ran into the problem that dynamic parameters do not work at all after I apply the plugin, even without using any sticky state.

**Comment by mgerstenblatt** [comment]
I ran into the same problem as @kuhnroyal -- dynamic parameters don't work at all in any state after applying the plugin.

**Comment by mgerstenblatt** [comment]
I've opened a PR in ui-router core that I think addresses this issue: https://github.com/ui-router/core/pull/72

**Comment by christopherthielen** [comment]
@mgerstenblatt I really appreciate the analysis and the PR!  I'm a bit afraid of side effects of changing the semantics for "stock ui-router" treeChanges.  It seems like sticky states itself could handle this by returning:

```
tc.reactivating = simulatedTC.to.slice(tc.retained.length);
```

do you want to try and let me know?

**Comment by mgerstenblatt** [comment]
@christopherthielen, I don't think that approach works because when you reload a state (after updating the parameters for example)  simulatedTC.to and tc.retained are the same length and you end up with an empty array.

I understand your concerns about modifying the ui-router core and would love a solution that didn't have to modify it. However, from scanning the usage of the treeChanges, I believe my changes have minimal impact.

**Comment by mgerstenblatt** [comment]
@christopherthielen, what about doing something similar where we apply the to params like in https://github.com/ui-router/core/blob/master/src/path/pathFactory.ts#L121

**Comment by mgerstenblatt** [comment]
@christopherthielen, something like this perhaps:

```
        function applyToParams(retainedNode, idx) {
            var cloned = _uirouter_core.PathNode.clone(retainedNode);
            cloned.paramValues = tc.to[idx].paramValues;
            return cloned;
        }
        if (shouldRewritePaths) {
            // The 'retained' nodes from the simulated transition's TreeChanges are the ones that will be reactivated.
            // (excluding the nodes that are in the original retained path)
            tc.reactivating = tc.retained.slice(tc.retained.length);
            // Entering nodes are the same as the simulated transition's entering
            tc.entering = simulatedTC.entering;
            // The simulatedTC 'exiting' nodes are inactives that are being exited because:
            // - The inactive state's params changed
            // - The inactive state is being reloaded
            // - The inactive state is a child of the to state
            tc.exiting = tc.exiting.concat(simulatedTC.exiting);
            // Rewrite the to path
            var retainedWithToParams = tc.retained.map(applyToParams);
            tc.to = retainedWithToParams.concat(tc.reactivating).concat(tc.entering);
        }
```

**Comment by mgerstenblatt** [comment]
@christopherthielen, something like https://github.com/ui-router/sticky-states/pull/15/files

**Comment by Aidamir** [comment]
I have faced with the issue while reactivating the sticky state with the updated dynamic parameter. (It did not update parameter at all). I did apply  @mgerstenblatt  patch #15 , but it still didn't  work. Then I found there is missing params update for reactivating  nodes  in patch. I have added one line to the code and it did start to work.  my updated code looks:
```

      // Rewrite the to path
      let retainedWithToParams = tc.retained.map((retainedNode,idx) => applyToParams(retainedNode,tc.to,idx));
      let reactivatingWithToParams = tc.reactivating.map((retainedNode,idx) => applyToParams(retainedNode,tc.to,(idx + tc.retained.length)));
      tc.to = retainedWithToParams.concat(reactivatingWithToParams).concat(tc.entering);
    }
```
@christopherthielen, Please explain why don't you accept #15 ?

**Comment by jnfeinstein** [comment]
I'm not seeing this issue specifically but am seeing lots of issues with query parameters.

@christopherthielen is there an ETA for fixes (this and #16)?  Would love to use this library.

**Comment by jnfeinstein** [comment]
Confirmed, this is the same issue I'm experiencing.

**Comment by nkoterba** [comment]
Have to agree with @vladtar here...

It looks like the PathFactory really should be returning `retainedWithParams` instead of `retained`.  It almost looks like the wrong variable was returned

**or** 

`retainedWithParams` should be included in the `treeChanges` object and then the **stickyStates** plugin can just utilize that instead of the original `retained` object.

**Comment by nerdo** [comment]
I believe this is the issue I ran into as well and didn't fully understand what was going on as I am new to sticky states. To be honest, I still don't *fully* understand why it happens, but I'll explain my scenario just in case it is slightly different.

Like the OP, I have two states set up in a similar fashion with few exceptions. First, my sticky state had parameters that were not declared dynamic. Secondly, my sticky state had no url, and I was not trying to change any parameters; I was trying to return to stateB with it's original parameters (essentially a "go back" function). 

Instead of everything in stateB showing up untouched, I found, through console logging, that the component that stateB was routed to was being re-initialized and it did seem like the parameters were not being passed through.

So there may be two issues here. The one that has been mentioned where the parameter is not changing, but it also seems like the sticky state is not recognizing that the state is already alive. If it did, I shouldn't be seeing the component get re-initialized.

I ended up writing a hacky workaround using a constant object to pass params and then injected that into the constant as a dependency.

For the small-scale project I ran into this on, this will do, but this is a pretty big deal breaker for use in larger scale projects.

**Comment by christopherthielen** [comment]
Looks like the change I made in `@uirouter/core` for this broke a lot of other things.  Stay tuned.

**Comment by mgerstenblatt** [comment]
@christopherthielen, for what it's worth, I've been using a patched version of sticky-states (https://github.com/ui-router/sticky-states/pull/15/files) instead of the change to ui-router/core in production since the summer and haven't had any issues.

**Comment by christopherthielen** [comment]
Fixed by 3696bf9f205a63e9fef1f6103284f95d4d54708f
